ct: Refactoring + end to end operation

Lazin opened the pull request:
This PR makes everything work end to end and separates data plane and control plane.

The data plane is `read_pipeline` and `write_pipeline` with all components. The PR changes them so they receive all metadata in the request (either `read_request` or `write_request`). The idea is that in the kafka layer we will fetch metadata from the corresponding location (either `dl_stm` or `cluster::partition`) and push it to the data plane as part of the read or write request.

This allows us to simplify the design. The CT app is just a service that moves data around in response to metadata. The metadata is stored in the partition.

In the `kafka` layer there is a `cloud_topic_partition` class. This is an implementation of the `partition_proxy` interface. This is a place where the metadata is fetched from the "control plane" and used to interact with the "data plane".


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64946-->
<!--640b879f-b3ae-46a7-acdb-bd69e8ebf518-->
<!--title-->
## Retry command for [Build#64946](https://buildkite.com/redpanda/redpanda/builds/64946)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves@{"cleanup_policy":"compact,delete"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--30cd0347-e479-48e8-8dd5-3c8b8843dbc9-->
<!--TEST RESULTS-->
<!--d3c9826f-3051-4b82-ab1c-2e6b0d028e9b-->
<!--TEST RESULTS-->
<!--2c614010-5856-482a-882b-8623f46deb37-->
<!--TEST RESULTS-->
<!--65219607-c075-41b3-be89-bbb4d79566ac-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64946">build#64946</a></summary>


| test_id                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64946#01966337-c78f-4686-be24-5d3a510f142c | FLAKY         | 18/21    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65989">build#65989</a></summary>


| test_class                 | test_method                                       | test_arguments                                                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:--------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterHealthOverviewTest  | cluster_health_overview_baseline_test             |                                                                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65989#0196cf78-938e-4431-9008-a0efa4ec4b45 | FLAKY         | 19/21    | upstream reliability is '96.02272727272727'. current run reliability is '90.47619047619048'. drift is 5.54654 and the allowed drift is set to 50. The test should PASS |
| NodePoolMigrationTest      | test_migrating_redpanda_nodes_to_new_pool         | {"balancing_mode": "off", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"}                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65989#0196cf78-938f-4003-aae7-7fcfac12c264 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                                                                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65989#0196cf78-9390-4137-b70e-5d569ce54d7d | FLAKY         | 17/21    | upstream reliability is '89.21739130434783'. current run reliability is '80.95238095238095'. drift is 8.26501 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65989#0196cf78-938e-4431-9008-a0efa4ec4b45 | FLAKY         | 20/21    | upstream reliability is '98.18181818181819'. current run reliability is '95.23809523809523'. drift is 2.94372 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65989#0196cf78-938e-4431-9008-a0efa4ec4b45 | FLAKY         | 20/21    | upstream reliability is '95.48989113530327'. current run reliability is '95.23809523809523'. drift is 0.2518 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66129">build#66129</a></summary>


| test_class                   | test_method                             | test_arguments                                                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:----------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"}                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66129#0196db9a-45e5-46a7-a9e4-b4403364423a | FLAKY         | 20/21    | upstream reliability is '97.87234042553192'. current run reliability is '95.23809523809523'. drift is 2.63425 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                    | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66129#0196db9a-45e3-4575-bafd-1cb77f1a2e38 | FLAKY         | 20/21    | upstream reliability is '97.72727272727273'. current run reliability is '95.23809523809523'. drift is 2.48918 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66137">build#66137</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66137#0196deb1-894e-4433-8589-665f9947ff06 | FLAKY         | 20/21    | upstream reliability is '97.09302325581395'. current run reliability is '95.23809523809523'. drift is 1.85493 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

Lazin committed:
ct: Add extent_meta struct

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update serializer to use extent_meta

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update write_request to use extent_meta

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update write_pipeline with extent_meta

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update aggregator

The cloud_topics/batcher/aggregator uses extent_meta now.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update batcher_test to use extent_meta

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update throttler_test

Use extent_meta instead of relying on placeholders directly

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update event_filter_test using extent_meta

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update cloud_topics/app

change return type of the 'write_and_debounce' to be 'extent_meta'

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
kafka: Update cloud_topic_partition

The CTP now sends data batches to the data plane and gets back list of
extent_meta values. After that it converts them to placeholder batches
and replicates them (actually a signle batch per produce request).

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update read_request

The read_request now uses extent_meta as an input.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use extent_meta in the placeholder_extent

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update placeholder_extent_reader

The reader now uses extent_meta as an input instead of the
model::record_batch.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update fetch_handler_test

Use extent_meta instead of the placeholder.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update 'materialize' interface

The method now takes a list of extnet_meta values instead of list of
placeholders.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update cloud_topic_partition

The conversion from placehodler baches to extent_meta values is
happening inside the cloud_topic_partition which acts as a coordinator
between the data plane and control plane.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add cloud_topics/api.h

This is a top level api of the data plane. It allows the caller to
either send L0 object to the cloud storage and get the metadata or
materialize the L0 object using the metadata.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use pimpl in cloud_topics/app

The 'app' now accepts a pointer to implementation (defined in the api.h)
in the c-tor. This is needed because the 'app' is created in the
'redpanda/application' and is a sharded service. And the sharded service
needs a concrete type, not the interface. So we can't create
'sharded_service<cloud_topics::api>' directly.

We also want to pass around the interface to avoid cyclic dependencies.
The 'app' implementation depends on 'cluster' because of the
'reconciler'. So passing 'app' to the cluster will create a cycle.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add cloud_topics/app_impl

This is the implementation of the 'cloud_topics/api'.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
application: Initialize cloud_topics_api

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
kafka: Add cloud_topic_partition

This commit adds a 'cloud_topic_partition' class which is an
implementation of the 'partiton_proxy'. The implementation connects the
control plane (cluster::partition and dl_stm) with the data plane
(cloud_topics::app).

The main goal here is to separate the control plane and data plane.
There is no explicit API boundary for the control plane. The
'cluster::partition' is queried directly. But it will be possible to
introduce the explicit control plane API in the future.

The data plane API boundary is defined by the 'cloud_topics::app'.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
kafka: Propagate cloud_topics data plane

...into the kafka/server. The reference to the 'cloud_topics::api'
should be eventually propagated to the 'cloud_topics_partition' which is
created in the read path and write path. In order to be able to do this
the reference to the data plane should be propagated to the kafka
server.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
kafka: Expose 'cloud_topics::api'

... through the request_context.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
kafka: Change signature of the make_parition_proxy

The function now accepts 'cloud_topics::api' pointer or a sharded
service. The implementation of the function creates either
'replicated_partition' or 'cloud_topic_partition' based on topic
configuration. If the partition belongs to cloud topic the second
parameter is used to initialize 'cloud_topic_partition' instance.
Otherwise it's unused.

All callsites are updated. Note that in some cases the dataplane is not
propagated to the subsystem yet (e.g. datalake, WASM) so these
subsystems will not work with cloud topics.

Both fetch and produce paths are updated and are invoking
'make_partition_proxy' using cloud topics api. This enables cloud topics
read and write paths so end to end operation is possible. Note that
delete records request handler is not using cloud topics api yet. This
will be fixed in future commits but for now delete records request will
not work with cloud topics.

There is no impact on normal topics/partitions.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update read_request

The read_request was accepting log_reader_config as an input. The
underlying pipeline was using it to contruct a log reader that fetches
placeholders from the underlying partition.

This commit changes the read_request so it accepts a set of placeholder
batches as an input. The idea is that at the higher level the
coordinator will fetch placeholders from the control plane (underlying
partition) and use them to construct the read request. In this scenario
the read_pipeline is merely a data layer. The input is a set of
coordinates that the data layer uses to fetch data.

Note that in this model the timequery is a metadata lookup which is
performed in the control plane. The result of the operation is a set of
placehodlers from which the read_request can be constructed.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Fix event subscription

Make sure that the subscription will send notification if the pipeline
already has some events that match the filter.

Instead of adding the filter to the list trigger it immediately.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update read_pipelne interface

The read_pipeline now accpets a dataplane_query as an input for the
make_reader method. The query is then passed to the c-tor of the
read_request.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Implement 'trigger_event' for the 'read_pipeline'

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Refactor read-path

Remove old read path which includes 'L0_read_path' and 'interfaces'.
The new code is located in 'read_path'. The difference is that the new
code doesn't have to query the underlying partition. So the interfaces
are no longer used and can be removed.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
fixup! update BUILD

Lazin committed:
fixup! read path update

Lazin committed:
ct: Refactor write_pipeline

Update the signature of the write_and_debounce and add trigger_event
method.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Improve background loop of the batcher

Add logging and start taking into account the return value of the
'run_once' method.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Fix batcher_test

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Refine :serializer deps

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Expose last_reconciled_offset via dl_stm

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Update reconciler

The reconciler now depends on cloud topics api. The CT partition can be
consumed using the cloud_topic_partition. But in order to create it the
api should be passed to the 'make_partition_proxy'.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add logging to throttler

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
